File: scripts/drama_team.py
#!/usr/bin/env python3
"""
短剧编剧团队 - 主控制脚本
使用多 Agent 协作方式创作短剧剧本

这是示例代码，需要根据 OpenClaw 的实际工具进行调整
"""

import os
import json
from pathlib import Path
from datetime import datetime

# Agent 提示词路径
PROMPTS_DIR = Path(__file__).parent.parent / "prompts"

def load_prompt(name):
    """加载 Agent 提示词"""
    prompt_file = PROMPTS_DIR / f"{name}.md"
    if prompt_file.exists():
        return prompt_file.read_text(encoding='utf-8')
    return f"[{name} prompt not found]"

# Agent 提示词
WRITER_PROMPT = load_prompt('writer')
ALIGNER_PROMPT = load_prompt('aligner')
RECORDER_PROMPT = load_prompt('recorder')

class DramaTeam:
    """短剧编剧团队 - 多 Agent 协作系统"""
    
    def __init__(self, project_dir):
        """
        初始化编剧团队
        
        Args:
            project_dir: 项目目录路径
        """
        self.project_dir = Path(project_dir)
        self.project_dir.mkdir(parents=True, exist_ok=True)
        
        # 初始化项目文件
        self._init_project_files()
        
    def _init_project_files(self):
        """初始化项目文件"""
        # 创建进度记录文件
        progress_file = self.project_dir / "script.progress.md"
        if not progress_file.exists():
            progress_file.write_text(f"""# 创作进度记录

## 项目信息
- **项目名称**: 未命名项目
- **开始日期**: {datetime.now().strftime('%Y-%m-%d')}
- **最后更新**: {datetime.now().strftime('%Y-%m-%d')}

## 创作进度
- **总集数**: 待定
- **已完成**: 0
- **进度**: 0%

## 创作决策
[等待创作开始]

## 修改历史
[等待创作开始]

## 伏笔追踪
[等待创作开始]
""", encoding='utf-8')
    
    def create_episode(self, episode_num, context, max_iterations=5):
        """
        创建单集剧本
        
        Args:
            episode_num: 集数
            context: 上下文信息（大纲、人物、集目录等）
            max_iterations: 最大迭代次数
            
        Returns:
            bool: 是否成功创建
        """
        iteration = 0
        
        while iteration < max_iterations:
            iteration += 1
            print(f"\n{'='*50}")
            print(f"EP-{episode_num:02d} - 第 {iteration} 次迭代")
            print(f"{'='*50}")
            
            # 1. 主编剧创作
            print("\n[1/3] 主编剧正在创作...")
            script = self._delegate_to_writer(
                goal=f"创作 EP-{episode_num:02d} 剧本",
                context=context,
                iteration=iteration
            )
            
            # 2. Script Aligner 审核
            print("[2/3] Script Aligner 正在审核...")
            result = self._delegate_to_aligner(
                goal=f"审核 EP-{episode_num:02d} 剧本",
                script=script
            )
            
            # 3. 检查审核结果
            if result['status'] == 'PASS':
                print("[3/3] ✅ 审核通过！")
                
                # Script Recorder 记录
                self._delegate_to_recorder(
                    goal=f"记录 EP-{episode_num:02d} 创作完成",
                    episode_num=episode_num,
                    context=context,
                    iterations=iteration
                )
                
                # 保存剧本
                self._save_script(episode_num, script)
                
                print(f"\n🎉 EP-{episode_num:02d} 创作完成！")
                return True
            else:
                print(f"[3/3] ❌ 审核未通过")
                print(f"反馈：{result['feedback'][:200]}...")
                
                # 添加反馈到上下文
                context['feedback'] = result['feedback']
                
                if iteration < max_iterations:
                    print(f"\n⏭️  继续第 {iteration + 1} 次迭代...")
        
        # 超过最大迭代次数
        print(f"\n⚠️  警告：EP-{episode_num:02d} 超过最大迭代次数 ({max_iterations})，需要人工介入")
        return False
    
    def _delegate_to_writer(self, goal, context, iteration=1):
        """
        委派任务给主编剧 Agent
        
        注意：这是示例代码，需要根据 OpenClaw 的实际工具进行调整
        """
        # 构建上下文
        full_context = f"""
{WRITER_PROMPT}

当前上下文：
{json.dumps(context, indent=2, ensure_ascii=False)}

迭代次数：{iteration}
"""
        if iteration > 1 and 'feedback' in context:
            full_context += f"\n上轮反馈:\n{context['feedback']}"
        
        # TODO: 使用 OpenClaw 的任务委派工具
        # result = delegate_task(
        #     goal=goal,
        #     context=full_context,
        #     toolsets=['file']
        # )
        # return result.output
        
        # 示例返回（实际使用时需要替换）
        return f"""
# EP-{episode_num:02d} [待填写集标题]

## 场景 1: [场景名称]
**时间**: [日/夜]
**地点**: [具体地点]

[这是示例剧本，实际使用时需要通过 delegate_task 调用 Writer Agent]
"""
    
    def _delegate_to_aligner(self, goal, script):
        """
        委派任务给 Script Aligner Agent
        
        注意：这是示例代码，需要根据 OpenClaw 的实际工具进行调整
        """
        full_context = f"""
{ALIGNER_PROMPT}

待审核剧本：
{script}
"""
        
        # TODO: 使用 OpenClaw 的任务委派工具
        # result = delegate_task(
        #     goal=goal,
        #     context=full_context,
        #     toolsets=['file']
        # )
        
        # 示例返回（实际使用时需要替换）
        # 实际应该调用 Aligner Agent 并解析返回结果
        return {
            'status': 'FAIL',
            'feedback': '这是示例反馈，实际使用时需要通过 delegate_task 调用 Aligner Agent'
        }
    
    def _delegate_to_recorder(self, goal, episode_num, context, iterations):
        """
        委派任务给 Script Recorder Agent
        
        注意：这是示例代码，需要根据 OpenClaw 的实际工具进行调整
        """
        full_context = f"""
{RECORDER_PROMPT}

当前进度：
- 集数：EP-{episode_num:02d}
- 迭代次数：{iterations}
- 项目信息：{json.dumps(context, indent=2, ensure_ascii=False)}

请更新 script.progress.md 文件
"""
        
        # TODO: 使用 OpenClaw 的任务委派工具
        # delegate_task(
        #     goal=goal,
        #     context=full_context,
        #     toolsets=['file']
        # )
        
        print("[3/3] Script Recorder 正在记录...")
    
    def _save_script(self, episode_num, script):
        """保存剧本到文件"""
        filename = f"EP-{episode_num:02d}.md"
        path = self.project_dir / filename
        
        with open(path, 'w', encoding='utf-8') as f:
            f.write(script)
        
        print(f"💾 剧本已保存：{path}")


def main():
    """主函数 - 示例用法"""
    print("=" * 60)
    print("短剧编剧团队 - 多 Agent 协作系统")
    print("=" * 60)
    
    # 创建项目
    project_dir = "my-drama-project"
    team = DramaTeam(project_dir)
    
    # 示例上下文
    context = {
        'outline': '复仇题材，女主被背叛后华丽转身',
        'characters': '女主：坚强独立；男主：霸道总裁',
        'episode_plan': 'EP-01: 女主回国，时尚周亮相'
    }
    
    # 创建第一集
    print("\n开始创作 EP-01...")
    success = team.create_episode(1, context)
    
    if success:
        print("\n" + "=" * 60)
        print("🎉 创作完成！")
        print("=" * 60)
    else:
        print("\n" + "=" * 60)
        print("⚠️  需要人工介入")
        print("=" * 60)


if __name__ == '__main__':
    main()

File: scripts/drama_team.js
#!/usr/bin/env node
/**
 * Drama Team - OpenClaw 多 Agent 协作实现
 * 
 * 使用 OpenClaw 的 sessions_spawn 启动独立 Agent
 */

const fs = require('fs');
const path = require('path');

class DramaTeam {
  constructor(projectPath) {
    this.projectPath = projectPath;
    this.skillsPath = path.join(__dirname, '..');
  }

  /**
   * 创作一集剧本（多 Agent 协作）
   * 
   * @param {number} epNum - 集数
   * @param {number} maxIterations - 最大迭代次数
   * @returns {Promise<boolean>} - 是否成功
   */
  async createEpisode(epNum, maxIterations = 5) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`开始创作 EP-${String(epNum).padStart(2, '0')}`);
    console.log('='.repeat(60));
    console.log('');

    // 读取上下文
    const context = this.loadContext();

    // 迭代创作
    for (let iteration = 1; iteration <= maxIterations; iteration++) {
      console.log(`\n${'='.repeat(50)}`);
      console.log(`EP-${String(epNum).padStart(2, '0')} - 第 ${iteration} 次迭代`);
      console.log('='.repeat(50));
      console.log('');

      // 1. Writer Agent 创作
      console.log('[1/3] Writer Agent 正在创作...');
      const script = await this.callWriterAgent(epNum, context, iteration);

      // 2. Aligner Agent 审核
      console.log('[2/3] Aligner Agent 正在审核...');
      const result = await this.callAlignerAgent(script);

      if (result.status === 'PASS') {
        console.log('[3/3] ✅ 审核通过！');

        // 3. Recorder Agent 记录
        console.log('\n[Recorder] 正在记录进度...');
        await this.callRecorderAgent(epNum, script, iteration);

        // 保存剧本
        this.saveScript(epNum, script);

        console.log(`\n✅ EP-${String(epNum).padStart(2, '0')} 创作完成！`);
        return true;
      } else {
        console.log('[3/3] ❌ 审核未通过');
        console.log(`\n反馈：${result.feedback}`);

        // 更新上下文（加入反馈）
        context.feedback = result.feedback;
        context.suggestions = result.suggestions || [];

        if (iteration < maxIterations) {
          console.log(`\n⏭️  继续第 ${iteration + 1} 次迭代...`);
        }
      }
    }

    console.log(`\n⚠️  达到最大迭代次数 (${maxIterations})，建议人工介入`);
    return false;
  }

  /**
   * 加载项目上下文
   */
  loadContext() {
    const context = {};

    const outlinePath = path.join(this.projectPath, 'outline.md');
    if (fs.existsSync(outlinePath)) {
      context.outline = fs.readFileSync(outlinePath, 'utf-8');
    }

    const characterPath = path.join(this.projectPath, 'character.md');
    if (fs.existsSync(characterPath)) {
      context.characters = fs.readFileSync(characterPath, 'utf-8');
    }

    const episodeIndexPath = path.join(this.projectPath, 'episode-index.md');
    if (fs.existsSync(episodeIndexPath)) {
      context.episodeIndex = fs.readFileSync(episodeIndexPath, 'utf-8');
    }

    return context;
  }

  /**
   * 调用 Writer Agent
   * 
   * 实际实现应该使用 sessions_spawn：
   * sessions_spawn({
   *   runtime: "acp",
   *   agentId: "drama-writer",
   *   task: "创作剧本",
   *   cwd: this.projectPath,
   *   attachments: [...]
   * })
   */
  async callWriterAgent(epNum, context, iteration) {
    const writerPrompt = fs.readFileSync(
      path.join(this.skillsPath, 'prompts/writer.md'),
      'utf-8'
    );

    const task = `
你现在是 Drama Team 的 Writer Agent。

请根据以下信息创作 EP-${String(epNum).padStart(2, '0')} 剧本：

## 项目上下文
- 大纲：见附件 outline.md
- 人物设定：见附件 character.md
- 集目录：见附件 episode-index.md

## 创作法则
${writerPrompt}

${iteration > 1 && context.feedback ? `
## 上次审核反馈
${context.feedback}

请根据反馈修改剧本。
` : ''}

## 输出要求
- 按照 templates/episode.md 格式输出
- 保存为 EP-${String(epNum).padStart(2, '0')}.md
`;

    // 实际调用 sessions_spawn
    // 这里返回任务描述，供实际使用
    return {
      task,
      agent: 'drama-writer',
      context
    };
  }

  /**
   * 调用 Aligner Agent
   * 
   * 实际实现应该使用 sessions_spawn
   */
  async callAlignerAgent(script) {
    const alignerPrompt = fs.readFileSync(
      path.join(this.skillsPath, 'prompts/aligner.md'),
      'utf-8'
    );

    const task = `
你现在是 Drama Team 的 Aligner Agent。

请审核以下剧本：

## 审核标准
${alignerPrompt}

## 待审核剧本
${script}

## 输出格式
返回 JSON：
{
  "status": "PASS 或 FAIL",
  "score": 4.5,
  "feedback": "具体反馈",
  "suggestions": ["修改建议1", "修改建议2"]
}
`;

    // 实际调用 sessions_spawn
    return {
      task,
      agent: 'drama-aligner',
      expectedOutput: {
        status: 'PASS 或 FAIL',
        score: 'number',
        feedback: 'string',
        suggestions: 'array'
      }
    };
  }

  /**
   * 调用 Recorder Agent
   */
  async callRecorderAgent(epNum, script, iterations) {
    const recorderPrompt = fs.readFileSync(
      path.join(this.skillsPath, 'prompts/recorder.md'),
      'utf-8'
    );

    const task = `
你现在是 Drama Team 的 Recorder Agent。

请记录以下创作进度：

## Recorder 职责
${recorderPrompt}

## 本次创作
- 集数：EP-${String(epNum).padStart(2, '0')}
- 迭代次数：${iterations}
- 状态：PASS

## 任务
更新 script.progress.md 文件
`;

    // 实际调用 sessions_spawn
    return {
      task,
      agent: 'drama-recorder'
    };
  }

  /**
   * 保存剧本
   */
  saveScript(epNum, script) {
    const scriptPath = path.join(
      this.projectPath,
      `EP-${String(epNum).padStart(2, '0')}.md`
    );
    console.log(`保存到：${scriptPath}`);
  }
}

// 导出
module.exports = DramaTeam;

// 命令行使用
if (require.main === module) {
  const projectPath = process.argv[2];
  const epNum = parseInt(process.argv[3]) || 1;

  if (!projectPath) {
    console.log('用法: node drama_team.js <项目路径> [集数]');
    console.log('示例: node drama_team.js my-drama 1');
    process.exit(1);
  }

  const team = new DramaTeam(projectPath);
  team.createEpisode(epNum).then(success => {
    process.exit(success ? 0 : 1);
  });
}
